tests/test_retries.py

```diff
@@ -0,0 +1,199 @@
+    items = []
+
+    with MockServer(EchoResource) as server:
+
+        class ItemPage(ItemWebPage):
+            def to_item(self):
+                raise Retry
+
+        class TestSpider(Spider):
+            name = "test_spider"
+            start_urls = [server.root_url]
+
+            custom_settings = {
+                "DOWNLOADER_MIDDLEWARES": {
+                    "scrapy_poet.InjectionMiddleware": 543,
+                },
+                "SPIDER_MIDDLEWARES": {
+                    "scrapy_poet.RetrySpiderMiddleware": 275,
+                },
+            }
+
+            def parse(self, response, page: ItemPage):
+                items.append(page.to_item())
+
+        crawler = make_crawler(TestSpider, {})
+        yield crawler.crawl()
+
+    assert items == []
+    assert crawler.stats.get_value("downloader/request_count") == 3
+    assert crawler.stats.get_value("retry/count") == 2
+    assert crawler.stats.get_value("retry/reason_count/page_object_retry") == 2
+    assert crawler.stats.get_value("retry/max_reached") == 1
+
+
+@inlineCallbacks
+def test_retry_max_configuration():
+    retries = deque([True, True, True, False])
+    items = []
+
+    with MockServer(EchoResource) as server:
+
+        class ItemPage(ItemWebPage):
+            def to_item(self):
+                if retries.popleft():
+                    raise Retry
+                return {"foo": "bar"}
+
+        class TestSpider(Spider):
+            name = "test_spider"
+            start_urls = [server.root_url]
+
+            custom_settings = {
+                "DOWNLOADER_MIDDLEWARES": {
+                    "scrapy_poet.InjectionMiddleware": 543,
+                },
+                "RETRY_TIMES": 3,
+                "SPIDER_MIDDLEWARES": {
+                    "scrapy_poet.RetrySpiderMiddleware": 275,
+                },
+            }
+
+            def parse(self, response, page: ItemPage):
+                items.append(page.to_item())
+
+        crawler = make_crawler(TestSpider, {})
+        yield crawler.crawl()
+
+    assert items == [{"foo": "bar"}]
+    assert crawler.stats.get_value("downloader/request_count") == 4
+    assert crawler.stats.get_value("retry/count") == 3
+    assert crawler.stats.get_value("retry/reason_count/page_object_retry") == 3
+    assert crawler.stats.get_value("retry/max_reached") is None
+
+
+@inlineCallbacks
+def test_non_retry_exception():
+    items = []
+
+    with MockServer(EchoResource) as server:
+
+        class ItemPage(ItemWebPage):
+            def to_item(self):
+                raise RuntimeError
+
+        class TestSpider(Spider):
+            name = "test_spider"
+            start_urls = [server.root_url]
+
+            custom_settings = {
+                "DOWNLOADER_MIDDLEWARES": {
+                    "scrapy_poet.InjectionMiddleware": 543,
+                },
+                "SPIDER_MIDDLEWARES": {
+                    "scrapy_poet.RetrySpiderMiddleware": 275,
+                },
+            }
```

Comment:
What do you think about having a base spider, with name and custom_settings, to reduce a bit code duplication in tests?